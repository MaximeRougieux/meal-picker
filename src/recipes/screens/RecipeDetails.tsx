import React from 'react';
import { ScrollView, Text, Image, View, StyleSheet } from 'react-native';
import { RecipeDetailsProps } from 'utils/navigation/types';
import { Meal } from 'recipes/types/APIMealsResponseData';

interface RecipeInstructionProps {
  item: string;
  index: number;
}

interface InstructionsProps {
  meal: Meal;
}

const RecipeInstruction = ({ item, index }: RecipeInstructionProps) => {
  return (
    <View style={styles.instructionContainer}>
      <Text>{`${index + 1}.   `}</Text>
      <Text>{item}</Text>
    </View>
  );
};

const Instructions = ({ meal }: InstructionsProps) => {
  const textInstructions = meal.strInstructions
    .split('\r\n')
    .filter(instruction => instruction);

  return (
    <>
      {textInstructions.map((item, index) => (
        <RecipeInstruction item={item} index={index} key={index} />
      ))}
    </>
  );
};
export const RecipeDetails = ({ route }: RecipeDetailsProps) => {
  const { meal } = route.params;
  return (
    <ScrollView>
      <Image source={{ uri: meal.strMealThumb }} style={styles.image} />
      <View style={styles.container}>
        <Text style={styles.title}>{meal.strMeal}</Text>
        <Text style={styles.category}>{meal.strCategory}</Text>
        <Instructions meal={meal} />
      </View>
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  image: {
    height: 300,
    width: '100%',
    borderBottomLeftRadius: 50,
    borderBottomRightRadius: 50,
  },
  category: {
    alignSelf: 'flex-end',
    fontSize: 20,
    fontStyle: 'italic',
  },
  container: {
    marginHorizontal: 12,
    marginTop: 16,
  },
  title: {
    color: 'black',
    fontSize: 36,
    fontFamily: 'Helvetica',
    fontWeight: '500',
    alignSelf: 'center',
    textAlign: 'center',
    marginBottom: 24,
  },
  instructionContainer: {
    flexDirection: 'row',
    margin: 8,
  },
});
