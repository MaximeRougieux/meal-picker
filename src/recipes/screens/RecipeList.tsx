import React, { useState } from 'react';
import {
  ActivityIndicator,
  ScrollView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import { RecipeListProps } from 'utils/navigation/types';
import { RecipeItem } from '../components/RecipeItem';
import SearchBar from '../components/SearchBar';
import useGetMeals from '../usecases/useGetMeals';

export const RecipeList = ({ navigation }: RecipeListProps) => {
  const [keyword, setKeyword] = useState('');
  const { loading, error, items } = useGetMeals(keyword);

  return (
    <View style={styles.main}>
      <Text style={styles.title}>Welcome to Meal Picker</Text>
      <SearchBar setKeyword={setKeyword} />
      <ScrollView>
        {loading ? (
          <ActivityIndicator testID="loading" />
        ) : error !== null ? (
          <Text style={styles.errorText}>{error}</Text>
        ) : items.length === 0 ? (
          <Text style={styles.errorText}>No meal available</Text>
        ) : (
          items.map(meal => (
            <TouchableOpacity
              key={meal.idMeal}
              onPress={() => navigation.navigate('Details', { meal: meal })}>
              <RecipeItem
                name={meal.strMeal}
                category={meal.strCategory}
                imageUrl={meal.strMealThumb}
              />
            </TouchableOpacity>
          ))
        )}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  main: {
    margin: 8,
  },
  errorText: {
    fontSize: 16,
    alignSelf: 'center',
  },
  title: {
    marginVertical: 16,
    fontSize: 36,
  },
});
