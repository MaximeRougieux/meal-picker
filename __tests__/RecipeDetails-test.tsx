import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';
import { RecipeDetails } from 'recipes/screens/RecipeDetails';

test('renders correctly', () => {
  const mockedMeal = {
    idMeal: '52929',
    strMeal: 'Timbits',
    strCategory: 'Dessert',
    strInstructions:
      'Sift together dry ingredients.\r\nMix together wet ingredients and incorporate into dry. Stir until smooth.\r\nDrop by teaspoonfuls(no bigger) into hot oil (365 degrees, no hotter), turning after a few moments until golden brown on all sides.\r\nRemove and drain.\r\nRoll in cinnamon sugar while still warm and serve.',
    strMealThumb:
      'https://www.themealdb.com/images/media/meals/txsupu1511815755.jpg',
  };
  const route: any = {
    params: { meal: mockedMeal },
  };
  const navigation: any = {};
  const { toJSON } = render(
    <RecipeDetails route={route} navigation={navigation} />,
  );
  expect(toJSON()).toMatchSnapshot();
});
