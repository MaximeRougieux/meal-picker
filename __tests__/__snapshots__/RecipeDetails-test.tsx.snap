// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView>
  <View>
    <Image
      source={
        Object {
          "uri": "https://www.themealdb.com/images/media/meals/txsupu1511815755.jpg",
        }
      }
      style={
        Object {
          "borderBottomLeftRadius": 50,
          "borderBottomRightRadius": 50,
          "height": 300,
          "width": "100%",
        }
      }
    />
    <View
      style={
        Object {
          "marginHorizontal": 12,
          "marginTop": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "fontFamily": "Helvetica",
            "fontSize": 36,
            "fontWeight": "500",
            "marginBottom": 24,
            "textAlign": "center",
          }
        }
      >
        Timbits
      </Text>
      <Text
        style={
          Object {
            "alignSelf": "flex-end",
            "fontSize": 20,
            "fontStyle": "italic",
          }
        }
      >
        Dessert
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "margin": 8,
          }
        }
      >
        <Text>
          0.   
        </Text>
        <Text>
          Sift together dry ingredients.
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "margin": 8,
          }
        }
      >
        <Text>
          1.   
        </Text>
        <Text>
          Mix together wet ingredients and incorporate into dry. Stir until smooth.
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "margin": 8,
          }
        }
      >
        <Text>
          2.   
        </Text>
        <Text>
          Drop by teaspoonfuls(no bigger) into hot oil (365 degrees, no hotter), turning after a few moments until golden brown on all sides.
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "margin": 8,
          }
        }
      >
        <Text>
          3.   
        </Text>
        <Text>
          Remove and drain.
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "margin": 8,
          }
        }
      >
        <Text>
          4.   
        </Text>
        <Text>
          Roll in cinnamon sugar while still warm and serve.
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
